### PR TITLE
Port codebase to Python3

### PR DESCRIPTION
**Pull request by shrianshChari**
Includes updating `print` statements ~~and replacing the [maintenance-only ujson](https://github.com/ultrajson/ultrajson?tab=readme-ov-file#project-status) with [orjson](https://pypi.org/project/orjson/)~~.

### CODE REVIEW

**Comment by monsanto** [comment]
When I did the PyPy conversion, I remember just using the `json` library directly. It seems like that's the case here too? `ujson` lines are commented out. So I don't know why we need another dependency at all, unless it has a proven speed-up.

File: TierUpdate9Doubles.py [hunk 1]
```diff
@@ -16,25 +16,25 @@ def getUsage(filename,col,weight,usage):
...
+	print(" + ---- + ------------------ + ------- + ")
+	print("[/CODE][/HIDE]")
 
 tiers = ['DUber','DOU', 'DUU', 'DNU']
 usageTiers = ['doublesou', 'doublesuu']
 
 def main(months):
 	file = open('keylookup.json', 'rb')
-	keyLookup = json.load(file)
+	keyLookup = json.loads(file.read())
```
Comment (monsanto) [inline TierUpdate9Doubles.py:37]: Change shouldn't be necessary if using the regular JSON library

File: batchLogReader.py [hunk 2]
```diff
@@ -557,17 +556,17 @@ def LogReader(filename,tier,movesets,ratings):
 							sys.stderr.write(str(nicks)+"\n")
 							return False
 						
-				move = line[7+5*spacelog+len(found):string.find(line,"|",7+5*spacelog+len(found))-1*spacelog]
+				move = line[7+5*spacelog+len(found): line.find("|", 7+5*spacelog+len(found)) - 1*spacelog]
```
Comment (monsanto) [inline batchLogReader.py:559]: Why did you rename this variable and change the formatting?

Comment (shrianshChari) [reply batchLogReader.py:559]: Format was changed to match how Python is generally written but I can revert it; function was changed because `find` and `rfind` are methods instead of functions in Python 3.

Comment (monsanto) [reply batchLogReader.py:559]: OK, I was confused because in Python 2 you'd do `line.find` anyway. But it looks like there is a `find` function on the string module, must be extremely outdated. If they have the same behavior then w/e

The line has a change anyway so probably not worth undoing. But in general for tricky stuff like this, it isn't the time to do other types of work

File: StatCounter.py [hunk 3]
```diff
@@ -353,7 +351,7 @@
 		nbins = len(histogram)
 
 		for start in range(len(stallCounter)):
-			if stallCounter[start] >= histogram[0][0]-binSize/2:
+			if stallCounter[start][0] >= histogram[0][0]-binSize/2:
```
Comment (monsanto) [inline StatCounter.py:354]: Explain?

File: batchLogReader.py [hunk 4]
```diff
@@ -275,8 +274,8 @@ def LogReader(filename,tier,movesets,ratings):
 			#	os.makedirs(d)
 			#msfile=open(outname,'ab')
 			if keyify(poke['species']) == 'meloettapirouette':
-				print filename
-			writeme={'trainer':trainer.encode('ascii', 'ignore'),
+				print(filename)
+			writeme={'trainer':trainer,
```
Comment (monsanto) [inline batchLogReader.py:278]: Explain? This means the output could potentially be in unicode; it's unclear if that breaks anything later

**Comment by monsanto** [comment]
> > When I did the PyPy conversion, I remember just using the `json` library directly. It seems like that's the case here too? `ujson` lines are commented out. So I don't know why we need another dependency at all, unless it has a proven speed-up.
> 
> `orjson` is faster than json according to online sources, but I guess we can use it later if it is necessary

Yes please, if we weren't using anything but the `json` library to begin with, this isn't the time for optimization

File: StatCounter.py [hunk 5]
```diff
@@ -246,10 +241,13 @@
...
-teammatefile.close()
-pickle.dump(encounterMatrix,encounterfile)
-encounterfile.close()
+filename="Raw/moveset/"+tier+"/teammate"+specs+".pickle"
+teammatefile=open(filename,'wb')
+with open(filename, 'wb') as teammate_file:
+    pickle.dump(teammateMatrix, teammate_file)
+filename="Raw/moveset/"+tier+"/encounterMatrix"+specs+".pickle"
+with open(filename, 'wb') as encounter_file:
+    pickle.dump(encounterMatrix, encounter_file)
```
Comment (monsanto) [inline StatCounter.py:250]: OK, fine, but not actually a change that is relevant to converting to Python 3

Comment (shrianshChari) [reply StatCounter.py:250]: The file context stuff isn't immediately relevant, but pickle files do have to be written to binary; before, they were just written as ASCII files I believe.

File: environment.yml [hunk 6]
```diff
@@ -1,7 +1,6 @@
-name: sus-env2
+name: sus-env3
```
Comment (monsanto) [inline environment.yml:1]: What is `environment.yml` for anyway? Is this still Conda garbage? I thought I got rid of that for PyPy

Did you need to use this to install? It would be nice if this worked without dependencies, but if we need dependencies, maybe moving to `poetry`, that's what I did last time I modernized Python code

Clarifying the installation would be good, since we'll have to change that when we resetup the server environment to run this with python 3

File: StatCounter.py (hunk 5 above)
Comment (monsanto) [reply StatCounter.py:250]: Ah OK. Fine as-is

**Comment by monsanto** [comment]
The rest looks OK to me. Thanks for taking this on

File: StatCounter.py (hunk 3 above)
Comment (shrianshChari) [reply StatCounter.py:354]: From Python 2:
```py
('stallCounter[start]', [-1.3699866305456414, 0.628122801542395])
('histogram[0][0]-binSize/2', -1.35)
('stallCounter[start] >= histogram[0][0]-binSize/2', True)
```
Essentially, Python 2 lets you make comparisons between a list and a number because it automatically gives you the 0-index number (or at least, I'm guessing that's what is happening). Python 3 is not happy with comparing a list and a number.

Comment (monsanto) [reply StatCounter.py:354]: It looks like a straight up bug in the Python 2 code.

File: environment.yml (hunk 6 above)
Comment (shrianshChari) [reply environment.yml:1]: I assumed you were using `conda` to get Python 2 running, but I can get PyPy set up. We will need dependencies because it uses `js2py` to grab some data from Pokemon Showdown.

Comment (monsanto) [reply environment.yml:1]: @Marty-D do you remember the deal with this? I feel like I got rid of this dependency at some point

Comment (monsanto) [reply environment.yml:1]: Maybe I didn't get rid of it. Either way I don't know how to use Conda. We should have instructions on how to use it in the README. We don't have to switch to something else in this PR if it works OK.

Comment (monsanto) [reply environment.yml:1]: Re: PyPy, my understanding is that it's not as useful with recent versions of Python now that it has it's own JIT. So just having it work in whatever configuration is used on the server is fine. I'm not sure my changes ever got used in the main workflow anyway.

File: StatCounter.py (hunk 3 above)
Comment (shrianshChari) [reply StatCounter.py:354]: Yet another reason why this port is happening.

Comment (monsanto) [reply StatCounter.py:354]: I have no way of verifying if [0] is the right fix here; it certainly seems similar to the other lines. Since this was broken to begin with, I'm not sure if we have to block on finding the right thing, but in that case, a comment that explains the situation seems warranted

Comment (shrianshChari) [reply StatCounter.py:354]: After doing some testing, it seems as though python 2 always treats lists as always being "greater" than numbers.
Test:

```py
>>> a = []
>>> b = 2
>>> a > b
True
>>> a = [1]
>>> a > b
True
>>> b = 0
>>> a = []
>>> a > b
True
>>> a = [-1]
>>> a > b
True
```

**Comment by monsanto** [comment]
If you could revert the `load -> loads + lead` type changes as well that would be good, it's unclear if this has the same performance characteristics at a glance (I don't know what these functions do internally; they might not buffer the entire whole thing into memory)

**Comment by monsanto** [comment]
And also a comment on the trainer name unicode stuff, what you had before was sensible and the obvious thing to do, so down the line it might not be clear to someone doing further cleanups why this is the way it is

File: environment.yml (hunk 6 above)
Comment (Marty-D) [reply environment.yml:1]: > @Marty-D do you remember the deal with this? I feel like I got rid of this dependency at some point

I'm not sure, I wanted to use your changes on the new sim server but couldn't find how to configure PyPy on nix, or if I even had the right permissions to do so. Didn't have time to waste getting sh or whoever to do it so I just used the old conda stuff, which I feel like ended up taking longer to figure out than in 2019.

```
mkdir -p ~/miniconda3
wget https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-x86_64.sh -O ~/miniconda3/miniconda.sh
bash ~/miniconda3/miniconda.sh -b -u -p ~/miniconda3
rm -rf ~/miniconda3/miniconda.sh
~/miniconda3/bin/conda init bash
git clone https://github.com/Marty-D/Smogon-Usage-Stats
cd Smogon-Usage-Stats/
conda env create
tmux attach
cd Smogon-Usage-Stats/
conda env create
cd ..
source ~/.bashrc
conda --version
cd Smogon-Usage-Stats/
tmux attach
  source activate sus-env2
  conda info --envs
    # conda environments:
    #
    base                     /home/marty/miniconda3
    sus-env2              *  /home/marty/miniconda3/envs/sus-env2
```

File: StatCounter.py (hunk 3 above)
Comment (monsanto) [reply StatCounter.py:354]: Okay, then one thing you could do is have your current line commented out, a comment saying what the deal is, and `if True:` with the rest. I don't know if the results of this script are deterministic, but if they are, it would allow us to run both Py2 and Py3 at the end of the month and see if there are any differences.

I don't know if that would be helpful or not. Just an idea.

File: environment.yml (hunk 6 above)
Comment (shrianshChari) [reply environment.yml:1]: > Maybe I didn't get rid of it. Either way I don't know how to use Conda. We should have instructions on how to use it in the README. We don't have to switch to something else in this PR if it works OK.

This can happen in a different PR then.

**Comment by shrianshChari** [comment]
> If you could revert the `load` -> `loads` + `read` type changes as well that would be good, it's unclear if this has the same performance characteristics at a glance (I don't know what these functions do internally; they might not buffer the entire whole thing into memory)

Did

**Comment by shrianshChari** [comment]
> And also a comment on the trainer name unicode stuff, what you had before was sensible and the obvious thing to do, so down the line it might not be clear to someone doing further cleanups why this is the way it is

Did

File: batchLogReader.py [hunk 7]
```diff
@@ -275,8 +273,9 @@ def LogReader(filename,tier,movesets,ratings):
 			#	os.makedirs(d)
 			#msfile=open(outname,'ab')
 			if keyify(poke['species']) == 'meloettapirouette':
-				print filename
-			writeme={'trainer':trainer.encode('ascii', 'ignore'),
+				print(filename)
+			# This sequence removes Unicode characters from trainer names.
```
Comment (monsanto) [inline batchLogReader.py:277]:
```suggestion
			# The Python 2 version of this code removed Unicode characters from the trainer names; it's unclear if this is required by downstream processing, but just in case preserve the behavior.
```

Your comment just restates what the code says; I was looking for the rationale behind it

**Comment by monsanto** [comment]
If it works locally for you then I approve these changes. I'll leave it unmerged until @Marty-D has time to integrate it in his workflow.

**Comment by shrianshChari** [comment]
> If it works locally for you then I approve these changes. Thanks again for the effort. I'll leave it unmerged until @Marty-D has time to integrate it in his workflow.

There is a bug in this code that I'm still looking through; I'll put more details once I'm back at my computer.

**Comment by shrianshChari** [comment]
To follow up on the bug, it's present in the Python 2 version as well, you can merge this PR.

**Comment by monsanto** [comment]
Well, what is the bug?

**Comment by shrianshChari** [comment]
In the synthetic data that I used to test the Tera types feature, I came across an issue where when `batchMovesetCounter.py` listed out the teammates, instead of there being an "Other" section, it would display the last teammate displayed and an extra pipe character. This is what it looked like:

```
 +----------------------------------------+ 
 | Dragapult                              | 
 +----------------------------------------+ 
 | Raw count: 1307                        | 
 | Avg. weight: ---                       | 
 | Viability Ceiling: 0                   | 
 +----------------------------------------+ 
 | Abilities                              | 
 | Clear Body 87.605%                     | 
 | Infiltrator 12.395%                    | 
 +----------------------------------------+ 
 | Items                                  | 
 | Choice Band 87.605%                    | 
 | Heavy-Duty Boots 12.395%               | 
 +----------------------------------------+ 
 | Spreads                                | 
 | Jolly:0/252/0/0/4/252 87.605%          | 
 | Timid:0/0/4/252/0/252 12.395%          | 
 +----------------------------------------+ 
 | Moves                                  | 
 | U-turn 100.000%                        | 
 | Tera Blast 87.605%                     | 
 | Dragon Darts 87.605%                   | 
 | Quick Attack 87.605%                   | 
 | Hex 12.395%                            | 
 | Will-O-Wisp 12.395%                    | 
 | Other 12.395%                          | 
 +----------------------------------------+ 
 | Teammates                              | 
 | Kingambit 103.039%                     | 
 | Primarina 90.840%                      | 
 | Samurott-Hisui 90.840%                 | 
 | Iron Crown 90.840%                     | 
 | Landorus-Therian 90.840%               | 
 | Zamazenta 12.199%                      | 
 | Zamazenta 12.199%                      | |
 +----------------------------------------+ 
 | Checks and Counters                    | 
 +----------------------------------------+
```

Specifically, the line that goes ` | Zamazenta 12.199%                      | |` is the bug.

I thought it appeared when porting to Python 3, but when I ran it in Python 2 with the same data, the same result appeared. I know that when the code runs the following (on line 235 of `batchMovesetCounter.py`:
```py
for i in range(len(table)): 
    if (total > .95 and x != 'Abilities') or (x == 'Abilities' and i>5) or (x == 'Spreads' and i>5) or (x == 'Teammates' and i>10) or (x == 'Checks and Counters' and i>11):
        if x == 'Moves':
            line = ' | %s %6.3f%%' % ('Other',400.0*(1.0-total))
        elif x not in ['Teammates','Checks and Counters']:
            line = ' | %s %6.3f%%' % ('Other',100.0*(1.0-total))
```

the `total` will be large enough to enter the conditional, but because `x = 'Teammates'`, it will just reuse the previous line when printing instead of printing the `'Other'` statistics. However, it seems to not occur with every Pokemon in my synthetic data (for example, this only occurs with a couple of Pokemon in my dataset) and I don't see it on [the Smogon stats](www.smogon.com/stats).

Tldr; fixing this (if it even is an issue) is better saved for another PR.

**Comment by monsanto** [comment]
Okay, it can be addressed later

File: environment.yml [hunk 8]
```diff
@@ -1,7 +1,6 @@
...
 dependencies:
 - ipython
-- python=2
-- ujson
-- pip:
-  - js2py
+- python=3.11
```
Comment (monsanto) [inline environment.yml:4]:
```suggestion
- python=3.12
```
Possible? No reason to go out of the gate with an outdated Python

**Comment by larry-the-table-guy** [comment]
In case it's helpful, here's how I had resolved the two merge conflicts on my local copy

<details><summary><code>batchMovesetCounter.py</code></summary>

```python
<<<<<<< HEAD
    species = keyLookup[filename[string.rfind(filename,'/')+1:]]
    species = reverseAliases.get(species, species)
    speciesKey = keyify(species)
=======
    species = keyLookup[filename[filename.rfind('/')+1:]]
    for alias in aliases:
        if species in aliases[alias]:
            species = alias
            break
>>>>>>> py3
```

(after)
```python
    species = keyLookup[filename[filename.rfind('/')+1:]]
    species = reverseAliases.get(species, species)
    speciesKey = keyify(species)
```

</details>

---

<details><summary><code>common.py</code></summary>

```python
<<<<<<< HEAD
import urllib2
import ujson as json
import re
=======
import urllib
import json
>>>>>>> py3

nonKeyRegex = re.compile('[^a-z0-9]+')
def keyify(s):
<<<<<<< HEAD
    return nonKeyRegex.sub('', s.lower())
=======
    sout = ''
    for c in s:
        if c in string.ascii_uppercase:
            sout = sout + c.lower()
        elif c in string.ascii_lowercase + '1234567890':
            sout = sout + c
    return sout
>>>>>>> py3
```

(after)
```python
import urllib
import json
import re

nonKeyRegex = re.compile('[^a-z0-9]+')
def keyify(s):
    return nonKeyRegex.sub('', s.lower())
```

</details>
